saged/models.py

```diff
@@ -134,3 +161,348 @@ def predict(self, dataset: np.array) -> np.array:
+        return self.model.predict(X)
+
+    def save_model(self, out_path: str) -> None:
+        """
+        Write the classifier to a file
+
+        Arguments
+        ---------
+        out_path: The path to the file to write the classifier to
+
+        Raises
+        ------
+        FileNotFoundError if out_path isn't openable
+        """
+
+        with open(out_path, 'wb') as out_file:
+            pickle.dump(self, out_file)
+
+    @classmethod
+    def load_model(classobject, model_path):
+        """
+        Read a pickeled model from a file and return it
+
+        Arguments
+        ---------
+        model_path: The location where the model is stored
+
+        Returns
+        -------
+        model: The model saved at `model_path`
+        """
+        with open(model_path, 'rb') as model_file:
+            return pickle.load(model_file)
+
+
+class ThreeLayerClassifier(nn.Module):
+    """ A basic three layer neural net for use in wrappers like FullyConnectedNet """
+    def __init__(self, model_params: dict):
+        super(ThreeLayerClassifier, self).__init__()
+        input_size = model_params['input_size']
+        output_size = model_params['output_size']
+
+        self.fc1 = nn.Linear(input_size, input_size // 2)
+        self.fc2 = nn.Linear(input_size // 2, input_size // 4)
+        self.fc3 = nn.Linear(input_size // 4, output_size)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+
+        return x
+
+
+class PytorchSupervised(ExpressionModel):
+    """
+    A wrapper class implementing the ExpressionModel API while remaining modular enough
+    to accept any supervised classifier implementing the nn.Module API
+    """
+    def __init__(self,
+                 config: dict,
+                 optimizer: torch.optim.Optimizer,
+                 loss_class: torch.nn.modules.loss._WeightedLoss,
+                 model_class: nn.Module) -> None:
+        """
+        Standard model init function. We use pass instead of raising a NotImplementedError
+        here in case inheriting classes decide to call `super()`
+
+        Arguments
+        ---------
+        config: The configuration information for the model
+        optimizer: The optimizer to be used when training the model
+        loss_class: The loss function class to use
+        model_class: The type of classifier to use
+        """
+        model_config = config['model']
+        self.config = config
+        self.model = model_class(model_config)
+        lr = float(model_config['lr'])
+        weight_decay = float(model_config.get('weight_decay', 0))
+        self.optimizer = optimizer(self.model.parameters(),
+                                   lr=lr,
+                                   weight_decay=weight_decay)
+
+        self.loss_class = loss_class
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```

Comment:
Do you always want to run PyTorch models on the GPU if one is available, or is there ever a case where you might want to override this and run things on the CPU?

I know in the past I've noticed shallow models (e.g. linear/logistic regression) in PyTorch can at times run faster on the CPU, but if you're only planning on using this class for multi-layer NNs (which is what it looks like here) you might be fine.

Reply:
That's a good point, it makes sense to put that in the config file too